src/assets/scripts/client/aircraft/AircraftModel.js

```diff
@@ -1387,15 +1439,51 @@ export default class AircraftModel {
                 return this._calculateLegalSpeed(this.model.speed.max);
 
             case MCP_MODE.SPEED.VNAV: {
-                const maxSpeed = this.mcp.speed;
-                const waypointSpeed = this.fms.currentWaypoint.speedRestriction;
-                const waypointHasSpeed = waypointSpeed !== -1;
+                // const mcpSpeed = this.mcp.speed;
+                // const waypointMaxSpeed = this.fms.currentWaypoint.speedMaximum;
+                //
+                // if (waypointMaxSpeed !== -1) {
+                //     return this._calculateLegalSpeed(waypointMaxSpeed);
+                // }
+                //
+                // return this._calculateLegalSpeed(mcpSpeed);
+                const hardRestrictedWaypointModel = this.fms.nextHardSpeedRestrictedWaypoint;
+
+                if (_isNil(hardRestrictedWaypointModel)) {
+                    return;
+                }
+
+                let waypointSpeed = hardRestrictedWaypointModel.speedMaximum;
+
+                if (waypointSpeed > this.speed) {
+                    const softOrHardRestrictedWaypoint = this.fms.nextSpeedRestrictedWaypoint;
```

Comment:
Abstracted contents of `AircraftModel._calculateTargetedSpeedVnav()` to:
- `AircraftModel._calculateTargetedSpeedVnavAcceleration()`
- `AircraftModel._calculateTargetedSpeedVnavDeceleration()`